src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KDF.java

```diff
@@ -0,0 +1,340 @@
+                    alg + "' failed.", e);
+        } finally {
+            if (p11SaltKey != null) {
+                p11SaltKey.releaseKeyID();
+            }
+            p11BaseKey.releaseKeyID();
+            token.releaseSession(session);
+        }
+    }
+
+    private P11Key.P11SecretKey convertKey(SecretKey key, String errorMessage) {
+        try {
+            return (P11Key.P11SecretKey) P11SecretKeyFactory.convertKey(token,
+                    key, null);
+        } catch (InvalidKeyException ike) {
+            throw new ProviderException(errorMessage, ike);
+        }
+    }
+
+    private abstract sealed class KeyMaterialMerger permits
+            AnyKeyMaterialMerger, KeyKeyMaterialMerger, DataKeyMaterialMerger {
+
+        final KeyMaterialMerger merge(SecretKey nextKeyMaterial) {
+            if (nextKeyMaterial instanceof SecretKeySpec) {
+                return merge(nextKeyMaterial.getEncoded());
+            } else {
+                return merge(convertKey(nextKeyMaterial,
+                        "Failure when merging key material."));
+            }
+        }
+
+        abstract SecretKey getKeyMaterial();
+
+        protected abstract KeyMaterialMerger merge(byte[] nextKeyMaterial);
+
+        protected abstract KeyMaterialMerger merge(
+                P11Key.P11SecretKey nextKeyMaterial);
+
+        protected final P11Key.P11SecretKey p11Merge(
+                P11Key.P11SecretKey baseKey, CK_MECHANISM ckMech,
+                int derivedKeyLen) {
+            Session session = null;
+            long baseKeyID = baseKey.getKeyID();
+            try {
+                session = token.getOpSession();
+                CK_ATTRIBUTE[] attrs = new CK_ATTRIBUTE[] {
+                        new CK_ATTRIBUTE(CKA_CLASS, CKO_SECRET_KEY),
+                        new CK_ATTRIBUTE(CKA_KEY_TYPE, CKK_GENERIC_SECRET),
+                };
+                long derivedKeyID = token.p11.C_DeriveKey(session.id(), ckMech,
+                        baseKeyID, attrs);
+                return (P11Key.P11SecretKey) P11Key.secretKey(session,
+                        derivedKeyID, "Generic", derivedKeyLen, null);
+            } catch (PKCS11Exception e) {
+                throw new ProviderException("Failure when merging key " +
+                        "material.", e);
+            } finally {
+                baseKey.releaseKeyID();
+                token.releaseSession(session);
+            }
+        }
+    }
+
+    private final class AnyKeyMaterialMerger extends KeyMaterialMerger {
+
+        protected KeyMaterialMerger merge(byte[] nextKeyMaterial) {
+            return P11KDF.this.new DataKeyMaterialMerger(nextKeyMaterial);
+        }
+
+        protected KeyMaterialMerger merge(P11Key.P11SecretKey nextKeyMaterial) {
+            return P11KDF.this.new KeyKeyMaterialMerger(nextKeyMaterial);
+        }
+
+        SecretKey getKeyMaterial() {
+            return null;
```

Comment:
If the ikms list is empty, this would return a null key.

Reply:
You're right, good catch!